### ISSUE TITLE
[BUG] Possible interaction between map tasks/k8s-array, inject-finalizers, ASG AZ Rebalancing and AWS Node Termination Handler 

### ISSUE THREAD

**Issue by CalvinLeather** [flyteorg/flyte#2788]
### Describe the bug

Okay this one is possibly multiple bugs that we are incorrectly lumping together (they are hard to reproduce, at-scale bugs with less clear error messages because they are low level/from K8s).

For a while now, in our production/at-scale deployment, when we scale up map tasks to decent scales (100s to low thousands of elements), we were getting these confusing errors (see below) about a pod getting deleted. We have --inject-finalizers option on the flyte propeller configmap, and were even able to confirm that finalizers were enabled via kubectl (`kubectl get pods --all-namespaces -o custom-columns=Kind:.kind,Name:.metadata.name,Finalizers:.metadata.finalizers`)

Despite this, we were getting errors about the pods being deleted manually etc. (when we hadn't intended to). We noticed that this was happening around the same time as dataplane logs indicating some API source was terminating the pods, but it took us a while to realize that the AWS Node Termination Handler was draining nodes (via taint and explicit requests to evict pods on them via K8s API).  We only realized this because we saw an ARN for the IAM role associated with the aws_managed lambda that does this: arn:aws:iam::042475242167:role/AWSWesleyClusterManagerLambda-NodeManagerRole-1TIDI5IYNBQSD

Once we realized this, we turned of AZ rebalancing on our ASG, and these issues appeared to stop.

So, we think there is some kind of issue with how the AWS Node Termination Handler is evicting pods, and how finalizers are used with flyte propeller to try to keep pod metadata around so flyte peopeller can poll for errrors. This bug has been hard to reproduce, but now that we know its about AZ Rebalancing and the AWS Node Termination Handler, we maybe can reproduce it at slightly lower scale.

**Error Messages from Console We Think Are All Similar or The Same Root Cause**
```
[0]: code:"Interrupted" message:"Pod failed. No message received from kubernetes.\r\n[fe6kensy-n2-0-0] terminated with exit code (137). Reason [Error]. Message: \n/home/ubuntu/miniconda/envs/pipeline/lib/python3.7/site-packages/requests/__init__.py:104: RequestsDependencyWarning: urllib3 (1.25.11) or chardet (5.0.0)/charset_normalizer (2.0.12) doesn't match a supported version!\n  RequestsDependencyWarning)\nNon-critical error: object passed to add_json_log_handler is not a RootLogger instance\n." 
```

```
[0]: code:"ResourceDeletedExternally" message:"resource not found, name [e2e-workflows-development/fexhijri-n2-0-0]. reason: pods \"fexhijri-n2-0-0\" not found" 
[2]: code:"ResourceDeletedExternally" message:"resource not found, name [e2e-workflows-development/fexhijri-n2-0-2]. reason: pods \"fexhijri-n2-0-2\" not found" 
[4]: code:"ResourceDeletedExternally" message:"resource not found, name [e2e-workflows-development/fexhijri-n2-0-4]. reason: pods \"fexhijri-n2-0-4\" not found" 
```

```
Workflow[ibd-statistics:development:ibd_statistics.workflows.ibd_statistics.ibd_statistics_wf] failed. RuntimeExecutionError: max number of system retry attempts [51/50] exhausted. Last known status message: failed at Node[n3]. RuntimeExecutionError: failed during plugin execution, caused by: failed to execute handle for plugin [k8s-array]: panic when executing a plugin [k8s-array]. Stack: [goroutine 729 [running]:
runtime/debug.Stack()
	/usr/local/go/src/runtime/debug/stack.go:24 +0x65
github.com/flyteorg/flytepropeller/pkg/controller/nodes/task.Handler.invokePlugin.func1.1()
	/go/src/github.com/flyteorg/flytepropeller/pkg/controller/nodes/task/handler.go:375 +0xfe
panic({0x1e51040, 0xc002560ff0})
	/usr/local/go/src/runtime/panic.go:838 +0x207
github.com/flyteorg/flytestdlib/bitarray.(*CompactArray).validateValue(...)
	/go/pkg/mod/github.com/flyteorg/flytestdlib@v1.0.4/bitarray/compact_array.go:34
github.com/flyteorg/flytestdlib/bitarray.(*CompactArray).SetItem(0xc0027d10c8, 0x2796db0?, 0xc0031d1320?)
	/go/pkg/mod/github.com/flyteorg/flytestdlib@v1.0.4/bitarray/compact_array.go:41 +0x2df
github.com/flyteorg/flyteplugins/go/tasks/plugins/array/k8s.LaunchAndCheckSubTasksState({0x2796db0, 0xc0031d1320}, {0x27a2700, 0xc0018f86e0?}, {0x7fae4518f8f0?, 0xc0005b8fc0}, 0x3966380, 0x0?, {0xc002b2b000, 0x77}, ...)
	/go/pkg/mod/github.com/flyteorg/flyteplugins@v1.0.8/go/tasks/plugins/array/k8s/management.go:243 +0x175a
github.com/flyteorg/flyteplugins/go/tasks/plugins/array/k8s.Executor.Handle({{0x7fae4518f8f0, 0xc0005b8fc0}, {{0x2789d50, 0xc000c544d0}}, {{0x2789d50, 0xc000c54580}}}, {0x2796db0, 0xc0031d1320}, {0x27a2700, 0xc0018f86e0})
	/go/pkg/mod/github.com/flyteorg/flyteplugins@v1.0.8/go/tasks/plugins/array/k8s/executor.go:116 +0x3f5
github.com/flyteorg/flytepropeller/pkg/controller/nodes/task.Handler.invokePlugin.func1(0x0?, {0x2796db0, 0xc0031d10b0}, {0x2799298?, 0xc0009923c0?}, 0x0?)
	/go/src/github.com/flyteorg/flytepropeller/pkg/controller/nodes/task/handler.go:382 +0x178
github.com/flyteorg/flytepropeller/pkg/controller/nodes/task.Handler.invokePlugin({{0x27970f8, 0xc0005c07e0}, {0x27848f0, 0xc0007d15e0}, 0xc00072fef0, 0xc00072ff20, 0xc00072ff50, {0x2798818, 0xc001516000}, 0xc0000ed600, ...}, ...)
	/go/src/github.com/flyteorg/flytepropeller/pkg/controller/nodes/task/handler.go:384 +0x9a
github.com/flyteorg/flytepropeller/pkg/controller/nodes/task.Handler.Handle({{0x27970f8, 0xc0005c07e0}, {0x27848f0, 0xc0007d15e0}, 0xc00072fef0, 0xc00072ff20, 0xc00072ff50, {0x2798818, 0xc001516000}, 0xc0000ed600, ...}, ...)
	/go/src/github.com/flyteorg/flytepropeller/pkg/controller/nodes/task/handler.go:617 +0x182b
github.com/flyteorg/flytepropeller/pkg/controller/nodes/dynamic.dynamicNodeTaskNodeHandler.handleParentNode({{0x279a148, 0xc00092c410}, {{0xc000d1bb40, {{...}, 0x0}, {0xc0003ca080, 0x4, 0x4}}, {0xc000d1bb60, {{...}, ...}, ...}, ...}, ...}, ...)
	/go/src/github.com/flyteorg/flytepropeller/pkg/controller/nodes/dynamic/handler.go:70 +0xd8
github.com/flyteorg/flytepropeller/pkg/controller/nodes/dynamic.dynamicNodeTaskNodeHandler.Handle({{0x279a148, 0xc00092c410}, {{0xc000d1bb40, {{...}, 0x0}, {0xc0003ca080, 0x4, 0x4}}, {0xc000d1bb60, {{...}, ...}, ...}, ...}, ...}, ...)
	/go/src/github.com/flyteorg/flytepropeller/pkg/controller/nodes/dynamic/handler.go:220 +0x9d0
github.com/flyteorg/flytepropeller/pkg/controller/nodes.(*nodeExecutor).execute(0xc000794e40, {0x2796db0, 0xc0031d0c00}, {0x2798698, 0xc0015e4500}, 0xc00035e3c0, {0x27ab1b8?, 0xc0029bf2b0?})
	/go/src/github.com/flyteorg/flytepropeller/pkg/controller/nodes/executor.go:382 +0x157
github.com/flyteorg/flytepropeller/pkg/controller/nodes.(*nodeExecutor).handleQueuedOrRunningNode(0xc000794e40, {0x2796db0, 0xc0031d0c00}, 0xc00035e3c0, {0x2798698?, 0xc0015e4500?})
	/go/src/github.com/flyteorg/flytepropeller/pkg/controller/nodes/executor.go:512 +0x227
github.com/flyteorg/flytepropeller/pkg/controller/nodes.(*nodeExecutor).handleNode(0xc000794e40, {0x2796db0, 0xc0031d0c00}, {0x277d258, 0xc000174000}, 0xc00035e3c0, {0x2798698?, 0xc0015e4500})
	/go/src/github.com/flyteorg/flytepropeller/pkg/controller/nodes/executor.go:736 +0x3c5
github.com/flyteorg/flytepropeller/pkg/controller/nodes.(*nodeExecutor).RecursiveNodeHandler(0xc000794e40, {0x2796db0, 0xc0031d07b0}, {0x27a63e8, 0xc002e9ae10}, {0x277d258, 0xc000174000}, {0x277d280?, 0xc000174000?}, {0x27a3810, ...})
	/go/src/github.com/flyteorg/flytepropeller/pkg/controller/nodes/executor.go:934 +0x705
github.com/flyteorg/flytepropeller/pkg/controller/nodes.(*nodeExecutor).handleDownstream(0x22f4f2d?, {0x2796db0, 0xc0031d07b0}, {0x27a63e8, 0xc002e9ae10}, {0x277d258, 0xc000174000?}, {0x277d280?, 0xc000174000}, {0x27a3810, ...})
	/go/src/github.com/flyteorg/flytepropeller/pkg/controller/nodes/executor.go:774 +0x3c5
github.com/flyteorg/flytepropeller/pkg/controller/nodes.(*nodeExecutor).RecursiveNodeHandler(0xc000794e40, {0x2796db0, 0xc0031d07b0}, {0x27a63e8, 0xc002e9ae10}, {0x277d258, 0xc000174000}, {0x277d280?, 0xc000174000?}, {0x27a3810, ...})
	/go/src/github.com/flyteorg/flytepropeller/pkg/controller/nodes/executor.go:941 +0x935
github.com/flyteorg/flytepropeller/pkg/controller/nodes.(*nodeExecutor).handleDownstream(0x22f4f2d?, {0x2796db0, 0xc0031d07b0}, {0x27a63e8, 0xc002e9ae10}, {0x277d258, 0xc000174000?}, {0x277d280?, 0xc000174000}, {0x27a3810, ...})
	/go/src/github.com/flyteorg/flytepropeller/pkg/controller/nodes/executor.go:774 +0x3c5
github.com/flyteorg/flytepropeller/pkg/controller/nodes.(*nodeExecutor).RecursiveNodeHandler(0xc000794e40, {0x2796db0, 0xc0031d07b0}, {0x27a63e8, 0xc002e9ae10}, {0x277d258, 0xc000174000}, {0x277d280?, 0xc000174000?}, {0x27a3810, ...})
	/go/src/github.com/flyteorg/flytepropeller/pkg/controller/nodes/executor.go:941 +0x935
github.com/flyteorg/flytepropeller/pkg/controller/workflow.(*workflowExecutor).handleRunningWorkflow(0xc0003ed2d0, {0x2796db0, 0xc0031d07b0}, 0xc000174000)
	/go/src/github.com/flyteorg/flytepropeller/pkg/controller/workflow/executor.go:147 +0x1b3
github.com/flyteorg/flytepropeller/pkg/controller/workflow.(*workflowExecutor).HandleFlyteWorkflow(0xc0003ed2d0, {0x2796db0, 0xc0031d07b0}, 0xc000174000)
	/go/src/github.com/flyteorg/flytepropeller/pkg/controller/workflow/executor.go:393 +0x40f
github.com/flyteorg/flytepropeller/pkg/controller.(*Propeller).TryMutateWorkflow.func2(0xc000f241b0, {0x2796db0, 0xc0031d07b0}, 0xc00234d848, 0x1e51040?)
	/go/src/github.com/flyteorg/flytepropeller/pkg/controller/handler.go:130 +0x18e
github.com/flyteorg/flytepropeller/pkg/controller.(*Propeller).TryMutateWorkflow(0xc000f241b0, {0x2796db0, 0xc0031d0030}, 0xc00272e000)
	/go/src/github.com/flyteorg/flytepropeller/pkg/controller/handler.go:131 +0x459
github.com/flyteorg/flytepropeller/pkg/controller.(*Propeller).Handle(0xc000f241b0, {0x2796db0, 0xc0031d0030}, {0xc002c86e70, 0x19}, {0xc002c86e8a, 0x8})
	/go/src/github.com/flyteorg/flytepropeller/pkg/controller/handler.go:205 +0x86d
github.com/flyteorg/flytepropeller/pkg/controller.(*WorkerPool).processNextWorkItem.func1(0xc001660d80, 0xc00234df28, {0x1e51040?, 0xc0007bcbd0})
	/go/src/github.com/flyteorg/flytepropeller/pkg/controller/workers.go:88 +0x510
github.com/flyteorg/flytepropeller/pkg/controller.(*WorkerPool).processNextWorkItem(0xc001660d80, {0x2796db0, 0xc0031d0030})
	/go/src/github.com/flyteorg/flytepropeller/pkg/controller/workers.go:99 +0xf1
github.com/flyteorg/flytepropeller/pkg/controller.(*WorkerPool).runWorker(0x2796db0?, {0x2796db0, 0xc001896870})
	/go/src/github.com/flyteorg/flytepropeller/pkg/controller/workers.go:115 +0xbd
github.com/flyteorg/flytepropeller/pkg/controller.(*WorkerPool).Run.func1()
	/go/src/github.com/flyteorg/flytepropeller/pkg/controller/workers.go:150 +0x59
created by github.com/flyteorg/flytepropeller/pkg/controller.(*WorkerPool).Run
	/go/src/github.com/flyteorg/flytepropeller/pkg/controller/workers.go:147 +0x285
]
```

### Expected behavior

The finalizers kep the pod around, preventing it from getting deleted (subsequently causing these errrors)

### Screenshots

Finalizers on:
<img width="977" alt="image" src="https://user-images.githubusercontent.com/6288302/185434180-15a1d411-9dc6-4edb-9a98-cc6628aeb605.png">

### Are you sure this issue hasn't been raised already?

- [X] Yes

### Have you read the Code of Conduct?

- [X] Yes

**Comment by CalvinLeather** [comment]
I will try to include more data from this later today (cloudwatch screenshots, timeline from an example failure etc.)

**Comment by CalvinLeather** [comment]
Moved here from https://flyte-org.slack.com/archives/CP2HDHKE1/p1658863230650909

**Comment by hamersaw** [comment]
@CalvinLeather - OK, I have tracked this down and think I know the fix, but don't have time to test it tonight. Basically, when the map task detects that a subtask Pod has been deleted externally it reports that as a system failure. We are tracking this using the `SystemFailures` bit array and it is incremented [here](https://github.com/flyteorg/flyteplugins/blob/cdfc191012c2b2d5e3d956dec031df790b75386e/go/tasks/plugins/array/k8s/management.go#L243) - coincides with the stack trace. The reason we track system errors specifically, is because when using interruptible tasks we can use the last retry to make the task non-interruptible in an attempt to disallow preemption. So when this bit array is initialized it uses the interruptible failure threshold as the max value [here](https://github.com/flyteorg/flyteplugins/blob/cdfc191012c2b2d5e3d956dec031df790b75386e/go/tasks/plugins/array/k8s/management.go#L116). The max value in the bit array dictates how many bits each item requires - so when we store values we check to make sure the value does not exceed it - which is happening in this case. Then because the number of system failures can be larger than the interruptible failure threshold we panic. This explains why this is only happening on the 2nd retry of the subtasks. We should be setting the max value the the maximum number of retry attempts rather than the interruptible failure threshold.

TL;DR - we need to initialize the system retries bit array [here](https://github.com/flyteorg/flyteplugins/blob/cdfc191012c2b2d5e3d956dec031df790b75386e/go/tasks/plugins/array/k8s/management.go#L116) with the max value `tCtx.TaskExecutionMetadata().GetMaxAttempts()`.

**Comment by hamersaw** [comment]
However, we should note that it seems like the region load-balancer is deleting Pods. So this will fix the panic in FlytePropeller during map task evaluations, but the Pods will still be deleted and if that happens larger than the map task maximum number of retries than the subtasks will fail and ultimately the map task will fail. It would be interesting to dive a little deeper.

**Comment by CalvinLeather** [comment]
That makes sense... I think the behaviour that load balancing/rebalancing activities causes retries up to the limit is reasonable/expected... just better to have the retries triggered than the panic that currently occurs. 

I wish there was some reasonable way to indicate to users why the pod went away on each retry (vs the "pod not found, manually deleted" error I think will happen once this panic is fixed), but I don't think there really is any good way to do this. As a human, to figure this out, I had to look at the K8s API logs to figure out what was draining/killing the nodes

**Comment by hamersaw** [comment]
> 

@CalvinLeather - do you recall if there was anything interesting in the `reason` or `message` fields in the pod status? So there are two cases where the "pod not found ... manually deleted" message occurs. either (1) the object is missing from k8s, in which case we can't really do anything because flyte parsing k8s logs is out of the question, or (2) the object has been marked as deleted by k8s, in this scenario we may be able to best effort parse the pod status and append the deletion message into the flyte error. not sure if this will work, but just a thought. might be worth it to open another issue to explore this?
